crates/bevy_render/src/render_resource/resource_macros.rs

```diff
@@ -0,0 +1,108 @@
+// structs containing wgpu types take a long time to compile. this is particularly bad for generic
+// structs containing wgpu structs. we avoid that in debug builds (and for cargo check and rust analyzer)
+// by boxing and type-erasing with the `render_resource_wrapper` macro.
+// analysis from https://github.com/bevyengine/bevy/pull/5950#issuecomment-1243473071 indicates this is
+// due to `evaluate_obligations`. we should check if this can be removed after a fix lands for
+// https://github.com/rust-lang/rust/issues/99188 (and after other `evaluate_obligations`-related changes).
+#[cfg(debug_assertions)]
+#[macro_export]
+macro_rules! render_resource_wrapper {
+    ($wrapper_type:ident, $wgpu_type:ty) => {
+        #[derive(Clone, Debug)]
+        pub struct $wrapper_type(Option<std::sync::Arc<Box<()>>>);
```

Comment:
I'm confused by the Option here. Why is this needed?

Reply:
it is so that we can `take` it in try_unwrap without having to do anything special for the `Drop` impl to work.

If we don't use an option, we have to `std::mem::forget` and/or `ManuallyDrop<>` (on `self` or `self.0` or both..?), or `std::mem::replace(self.0)` (which creates a new `Arc<Box<()>>` just to drop it again in `Drop`). 

I can do one of those if you prefer, I will need to dig around to get comfortable with it, but since this is `#[cfg(debug_assertions)]` code perhaps it's not necessary?

Reply:
Ok that makes sense to me. We could solve the `try_unwrap` problem with another layer of wrapper types (so we can implement drop on the internal type, making it possible to move out on try_unwrap()). But implementing drop for that internal type is still problematic.